### ISSUE TITLE
How to access metadata information like MaxLength ?

### ISSUE THREAD

**Issue by francisharvey** [OData/odata.net#2066]
Here's an extract of the metadata file for the D365 Business Central Item entity.

`<EntityType Name="item">
        <Key>
          <PropertyRef Name="id" />
        </Key>
        <Property Name="id" Type="Edm.Guid" Nullable="false" />
        <Property Name="number" Type="Edm.String" MaxLength="20" />
        <Property Name="displayName" Type="Edm.String" MaxLength="100" />
        <Property Name="type" Type="Microsoft.NAV.itemType" />
        <Property Name="blocked" Type="Edm.Boolean" />
        <Property Name="gtin" Type="Edm.String" MaxLength="14" />
        <Property Name="unitPrice" Type="Edm.Decimal" Scale="Variable" />`

As of now, Is there a way to access the attributes of fields?
I think it would be useful to have MaxLength in the entity type property attributes like the OriginalNameAttribute, not sure for the need of the others...

Thanks,
Francis Harvey

**Comment by habbes** [comment]
Hi @francisharvey  you can get an edm type reference for the specific property type you want to extract metadata from. In the case of a string property, the `IEdmStringTypeReference` has properties like `MaxLength`. Here's how you could extract such information from an `EdmModel`:

```c#
IEdmEntityType entityType = model.FindType("item") as IEdmEntityType;
IEdmProperty property = entityType.FindProperty("displayName");
IEdmStringTypeReference propertyRef = property.Type.AsString();
int? maxLength = propertyRef.MaxLength;
```

**Comment by francisharvey** [comment]
Hi @habbes, 

How do you get the model from the Context?

Let's say I have a a D365 Finance and Operation metadata file that is 23mb...
I don't want to download it again, or create a temporary model from it, every times I need to know a field length...

Since I have the entity type and know the property name and sometime still need to verify the OriginalNameAttribute (for recursive expand with filter), I was under the impression adding an attribute would be easier to implement and use.

**Comment by habbes** [comment]
@francisharvey could you provide more information about how you're using OData? I assumed you had an `IEdmModel` object representing your schema. Are you referring to the `DataServiceContext` from the OData Client library? Did you create your client using OData Connected Service?

**Comment by francisharvey** [comment]
Hi @habbes, yes the DataServiceContext classes are generated using the OData Connected Service. I'm looking for a way to validate entity field values before sending them to the server, because request fails when the field lengths are wrong.

**Comment by habbes** [comment]
Hi @francisharvey I don't think the `DataServiceContext` exposes the underlying `IEdmModel`. But the code generated by OCS does read and parse the edm xml file when the context is instantiated. And it also happens to set hook that loads this parsed model in the `LoadServiceModel` property of the `DataServiceContext.Format` property. So I think you can use that hook as a work around to get the `IEdmModel`:

```c#
IEdmModel model = context.Format.LoadServiceModel();
```
Then you can use the `model` to access the metadata of the properties you're interested in.

This is not ideal, and might not work in all cases, but it should work for clients generated by OData Connected Service.

Let me know whether this works for you.

**Comment by francisharvey** [comment]
Hi @habbes,

Thank you for this workaround.

While it is highly inefficient, it may work for some.
But in an automated scenario, it may also not be usable at all. 
I mean without human writing code and trying strings thing until it works...

Here's what I got :

Dim itemObject As New MyTenant.Sandbox.Item
Dim Itemtype As Type = itemObject.GetType()

Dim model As Microsoft.OData.Edm.IEdmModel = Context.Format.LoadServiceModel().Invoke()
Dim t1 = model.FindDeclaredType("Item")
Dim t2 = model.FindDeclaredType("MyTenant.Sandbox.Item")
Dim t3 = model.FindType("Item")
Dim t4 = model.FindType("MyTenant.Sandbox.Item")
Dim t5 = model.FindType("item")
Dim t6 = model.FindDeclaredType("item")

Dim t7 = model.FindDeclaredType(Itemtype.Name)
Dim t8 = model.FindType(Itemtype.FullName)

Dim t9 = model.FindDeclaredType("Microsoft.NAV.item") ' => Not present anywhere in generated classes (Only in schema element `<Schema xmlns = "http://docs.oasis-open.org/odata/ns/edm" Namespace="Microsoft.NAV">`)

Dim t10 = model.FindType("Microsoft.NAV.item")

If Not IsNothing(t10) Then ' only t9 / t10 are working
    Dim entityType As Microsoft.OData.Edm.IEdmEntityType = t10

    Dim propInfo As Reflection.PropertyInfo = Itemtype.GetProperty("DisplayName")

    Dim prop1 = entityType.FindProperty("DisplayName")
    Dim prop2 = entityType.FindProperty(propInfo.Name)

    Dim prop3 = entityType.FindProperty("displayName")

    Dim originalNameAttr As Microsoft.OData.Client.OriginalNameAttribute = propInfo.GetCustomAttribute(GetType(Microsoft.OData.Client.OriginalNameAttribute))
    Dim originalName As String = originalNameAttr.OriginalName
    Dim prop4 = entityType.FindProperty(originalName)

    Dim maxLen As Integer? = Nothing

    If Not IsNothing(prop4) Then ' only prop3 / prop4 are working
        If prop4.Type.IsString Then
            Dim stringProp = prop4.Type.AsString()
            maxLen = stringProp.MaxLength
        End If
    End If

    If maxLen.HasValue Then
        maxLen.Value.ToString()
    End If
End If

The names to use has to be those in the schema, with the schema namespace, not even the one we specified during the generation, and property name must also be the schema ones, the originalName...

This is why I'm suggesting again why not create and use a customAttribute like the OriginalNameAttribute?

It would be usable directly from generated classes, processed only one time while generating classes and as simple as this to use on the client side:

Dim maxLen As Integer? = Nothing

Dim itemObject As New MyTenant.Sandbox.Item
Dim Itemtype As Type = itemObject.GetType()
Dim propInfo As Reflection.PropertyInfo = Itemtype.GetProperty("DisplayName")

Dim maxLengthAttr As Microsoft.OData.Client.MaxLengthAttribute = propInfo.GetCustomAttribute(GetType(Microsoft.OData.Client.MaxLengthAttribute))
If Not IsNothing(maxLengthAttr) Then
        maxLen = maxLengthAttr.MaxLength
End If

Let me know what you think about it,
Francis Harvey

Sorry about the code tags, I tried but the editor did not let me use them...
